File: src/bit_rot/scanner.py
"""Scanner module for directory traversal and file integrity verification."""

import logging
import os
import threading
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Optional

from .database import Database, FileRecord
from .hasher import Hasher

logger = logging.getLogger(__name__)


@dataclass
class SyncResult:
    """Results from a sync operation."""

    files_scanned: int
    files_added: int
    files_modified: int
    files_moved: int
    files_removed: int
    errors: list[str]


@dataclass
class ScrubResult:
    """Results from a scrub operation."""

    files_validated: int
    files_corrupted: list[str]
    errors: list[str]


class Scanner:
    """Directory scanner with move detection and scrubbing capabilities."""

    def __init__(self, hasher: Hasher, stop_event: Optional[threading.Event] = None):
        """Initialize scanner.

        Args:
            hasher: Hasher instance for computing file hashes
            stop_event: Event to signal cancellation
        """
        self.hasher = hasher
        self.stop_event = stop_event

    @staticmethod
    def check_canary(root_path: Path) -> bool:
        """Verify that the canary file exists.

        Args:
            root_path: Root directory to check

        Returns:
            True if canary exists, False otherwise
        """
        canary_path = root_path / ".bitrot-canary"
        exists = canary_path.exists()

        if exists:
            logger.info(f"Canary check passed: {canary_path}")
        else:
            logger.critical(f"Canary check FAILED: {canary_path} not found")

        return exists

    @staticmethod
    def _ensure_long_path(path: Path) -> Path:
        """Ensure path uses long path prefix on Windows to bypass 260 char limit.

        Args:
            path: Path to check

        Returns:
            Path with long path prefix if on Windows
        """
        if os.name == "nt":
            abs_path = str(path.resolve())
            if not abs_path.startswith("\\\\?\\"):
                return Path(f"\\\\?\\{abs_path}")
        return path

    def sync_directory(self, root_path: Path, db: Database) -> SyncResult:
        """Sync directory with database (Phases 1-3).

        Phase 1: Walk directory and record file metadata
        Phase 2: Detect moved files
        Phase 3: Detect deleted files

        Args:
            root_path: Root directory to scan
            db: Database instance

        Returns:
            SyncResult with operation statistics
        """
        # Ensure we can handle long paths on Windows
        root_path = self._ensure_long_path(root_path)

        logger.info(f"Starting sync operation for: {root_path}")
        session_start = datetime.now().isoformat()

        files_scanned = 0
        files_added = 0
        files_modified = 0
        files_moved = 0
        files_removed = 0
        errors = []

        # Get existing files from database
        db_files = db.get_all_files()
        logger.info(f"Loaded {len(db_files)} existing files from database")

        # Phase 1: Walk directory tree (don't follow symlinks to avoid loops)
        logger.info("Phase 1: Walking directory tree")
        current_files: dict[str, tuple[int, float]] = {}  # path -> (size, mtime)

        # System folders to skip (Windows-specific)
        system_folders = {
            "$RECYCLE.BIN",
            "System Volume Information",
            "Windows",
            "Program Files",
            "Program Files (x86)",
            "ProgramData",
            "Recovery",
            "$Recycle.Bin",  # Case variation
        }

        for root, dirs, files in os.walk(root_path, followlinks=False):
            if self.stop_event and self.stop_event.is_set():
                logger.info("Scanning interrupted")
                break

            # Skip system folders (modify dirs in-place to prevent descending)
            dirs[:] = [d for d in dirs if d not in system_folders]

            # Skip the database file itself if it's in the current directory
            if "bitrot.db" in files:
                files.remove("bitrot.db")

            for filename in files:
                if self.stop_event and self.stop_event.is_set():
                    break
                filepath = Path(root) / filename
                try:
                    stat = filepath.stat()
                    current_files[str(filepath.absolute())] = (
                        stat.st_size,
                        stat.st_mtime,
                    )
                    files_scanned += 1

                    if files_scanned % 1000 == 0:
                        logger.info(f"Scanned {files_scanned} files...")

                except (PermissionError, OSError) as e:
                    error_msg = f"Error accessing {filepath}: {e}"
                    logger.warning(f"{error_msg}")
                    errors.append(error_msg)

        logger.info(f"Phase 1 complete: Scanned {files_scanned} files")

        # Phase 2: Process files and detect moves
        logger.info("Phase 2: Processing files and detecting moves")

        # Track which DB files we've seen
        seen_db_paths = set()

        for current_path, (size, mtime) in current_files.items():
            if self.stop_event and self.stop_event.is_set():
                logger.info("Scanning interrupted (Phase 2)")
                break

            if current_path in db_files:
                # File exists in same location - check if modified
                record = db_files[current_path]

                # Compare size and mtime to detect modifications
                size_changed = size != record.file_size
                mtime_changed = abs(mtime - record.mtime) > 0.001  # 1ms tolerance

                if size_changed or mtime_changed:
                    # File was modified - re-compute hash
                    try:
                        logger.info(
                            f"Detected modification: {current_path} "
                            f"(size: {record.file_size} -> {size}, "
                            f"mtime changed: {mtime_changed})"
                        )
                        new_hash = self.hasher.compute_hash(Path(current_path))
                        db.stage_file_update(
                            abs_path=current_path,
                            hash=new_hash,
                            file_size=size,
                            mtime=mtime,
                            added_at=record.added_at,
                            last_scrubbed=None,  # Reset scrub status on modification
                            scrub_count=0,  # Reset scrub count on modification
                        )
                        files_modified += 1
                    except (PermissionError, OSError) as e:
                        error_msg = f"Error hashing modified file {current_path}: {e}"
                        logger.warning(f"{error_msg}")
                        errors.append(error_msg)
                else:
                    # File unchanged - just update last_seen
                    db.stage_file_update(
                        abs_path=current_path,
                        hash=record.hash,
                        file_size=size,
                        mtime=mtime,
                        added_at=record.added_at,
                        last_scrubbed=record.last_scrubbed,
                        scrub_count=record.scrub_count,
                    )

                seen_db_paths.add(current_path)
            else:
                # New file or potentially moved file
                # Check if a file with matching size/mtime exists in DB at different path
                moved_from = self._find_moved_file(
                    current_path, size, mtime, db_files, seen_db_paths, current_files
                )

                if moved_from:
                    # Size and mtime already matched in _find_moved_file
                    # No need to rehash - the file is uniquely identified
                    old_record = db_files[moved_from]
                    logger.info(f"Detected move: {moved_from} -> {current_path}")
                    db.stage_file_update(
                        abs_path=current_path,
                        hash=old_record.hash,  # Reuse existing hash
                        file_size=size,
                        mtime=mtime,
                        added_at=old_record.added_at,
                        last_scrubbed=old_record.last_scrubbed,
                        scrub_count=old_record.scrub_count,
                    )
                    # Mark old path for removal
                    db.stage_file_removal(moved_from)
                    seen_db_paths.add(moved_from)
                    files_moved += 1
                else:
                    # New file - compute hash and add
                    try:
                        file_hash = self.hasher.compute_hash(Path(current_path))
                        self._add_new_file(current_path, file_hash, size, mtime, db)
                        files_added += 1

                        if files_added % 100 == 0:
                            logger.info(f"Added {files_added} new files...")

                    except (PermissionError, OSError) as e:
                        error_msg = f"Error hashing {current_path}: {e}"
                        logger.warning(f"{error_msg}")
                        errors.append(error_msg)

        logger.info(
            f"Phase 2 complete: Added {files_added} files, "
            f"Modified {files_modified} files, Moved {files_moved} files"
        )

        # Phase 3: Detect deleted files
        logger.info("Phase 3: Detecting deleted files")

        for db_path in db_files:
            if self.stop_event and self.stop_event.is_set():
                break
            if db_path not in current_files and db_path not in seen_db_paths:
                logger.info(f"Detected deletion: {db_path}")
                db.stage_file_removal(db_path)
                files_removed += 1

        logger.info(f"Phase 3 complete: Removed {files_removed} files")

        result = SyncResult(
            files_scanned=files_scanned,
            files_added=files_added,
            files_modified=files_modified,
            files_moved=files_moved,
            files_removed=files_removed,
            errors=errors,
        )

        logger.info(
            f"Sync complete - Scanned: {files_scanned}, "
            f"Added: {files_added}, Modified: {files_modified}, Moved: {files_moved}, "
            f"Removed: {files_removed}, Errors: {len(errors)}"
        )

        return result

    def _find_moved_file(
        self,
        current_path: str,
        size: int,
        mtime: float,
        db_files: dict[str, FileRecord],
        seen_paths: set[str],
        current_files: dict[str, tuple[int, float]],
    ) -> Optional[str]:
        """Find a file in the database with matching size and mtime.

        IMPORTANT: Only returns a match if the old location no longer exists
        in the current scan. This prevents false move detection when duplicate
        files exist (e.g., two identical files, moving one shouldn't be detected
        as a move if the original still exists).

        Args:
            current_path: Current path of the file
            size: File size in bytes
            mtime: File modification time
            db_files: Dictionary of database files
            seen_paths: Set of paths already processed
            current_files: Dictionary of current files from scan

        Returns:
            Path of the matching file in DB, or None if not found
        """
        for db_path, record in db_files.items():
            if self.stop_event and self.stop_event.is_set():
                break

            if db_path in seen_paths:
                continue

            # CRITICAL: Only consider it a move if old location no longer exists
            if db_path in current_files:
                # Old location still exists - this is a duplicate, not a move
                continue

            # Check if size matches and path is different
            if record.file_size == size and db_path != current_path:
                # Check mtime (allow 1 second tolerance for filesystem differences)
                try:
                    db_mtime = datetime.fromisoformat(record.last_seen).timestamp()
                    if abs(db_mtime - mtime) < 1.0:
                        return db_path
                except (ValueError, OSError):
                    continue

        return None

    def _add_new_file(
        self, path: str, file_hash: str, size: int, mtime: float, db: Database
    ) -> None:
        """Add a new file to the database.

        Args:
            path: Absolute path to the file
            file_hash: BLAKE3 hash of the file
            size: File size in bytes
            mtime: File modification time
            db: Database instance
        """
        logger.debug(f"Adding new file: {path}")
        db.stage_file_update(abs_path=path, hash=file_hash, file_size=size, mtime=mtime)

    def scrub_files(
        self, db: Database, percentage: float, frequency: str
    ) -> ScrubResult:
        """Scrub files to detect bit rot (Phase 4).

        Args:
            db: Database instance
            percentage: Percentage of files to scrub (0.1 to 100.0)
            frequency: Scrub frequency (daily/weekly/monthly)

        Returns:
            ScrubResult with validation statistics
        """
        logger.info(f"Starting scrub operation ({percentage}%, {frequency} frequency)")

        # Determine minimum age based on frequency
        min_age_days = None
        if frequency == "weekly":
            min_age_days = 7
        elif frequency == "monthly":
            min_age_days = 30

        # Get files to scrub
        files_to_scrub = db.get_files_for_scrub(percentage, min_age_days)
        logger.info(f"Selected {len(files_to_scrub)} files for scrubbing")

        files_validated = 0
        files_corrupted = []
        errors = []

        for i, record in enumerate(files_to_scrub, 1):
            if self.stop_event and self.stop_event.is_set():
                logger.info("Scrubbing interrupted")
                break

            try:
                filepath = Path(record.abs_path)

                # Check if file still exists
                if not filepath.exists():
                    logger.warning(f"File no longer exists: {filepath}")
                    errors.append(f"File not found: {filepath}")
                    continue

                # Compute current hash
                current_hash = self.hasher.compute_hash(filepath)

                # Compare with stored hash
                if current_hash == record.hash:
                    # Hash matches - file is intact
                    db.update_scrub_status(record.abs_path)
                    files_validated += 1

                    if files_validated % 100 == 0:
                        logger.info(
                            f"Validated {files_validated}/{len(files_to_scrub)} files..."
                        )
                else:
                    # Hash mismatch - BIT ROT DETECTED!
                    error_msg = (
                        f"BIT ROT DETECTED: {filepath}\n"
                        f"  Expected: {record.hash}\n"
                        f"  Got:      {current_hash}"
                    )
                    files_corrupted.append(record.abs_path)

            except (PermissionError, OSError) as e:
                error_msg = f"Error scrubbing {record.abs_path}: {e}"
                logger.warning(f"{error_msg}")
                errors.append(error_msg)

        result = ScrubResult(
            files_validated=files_validated,
            files_corrupted=files_corrupted,
            errors=errors,
        )

        logger.info(
            f"Scrub complete - Validated: {files_validated}, "
            f"Corrupted: {len(files_corrupted)}, Errors: {len(errors)}"
        )

        return result

File: internal/scanner/scanner.go
// Package scanner implements directory walking, file-integrity sync, and
// periodic scrubbing using a producer-consumer worker pool.
//
// # Sync
//
// A sync operation walks an entire directory tree and, for each regular file,
// determines whether the file is new, modified, moved, or deleted relative to
// the contents of the integrity database.  Only files whose size or mtime
// differ from the stored values are re-hashed; unchanged files are recorded as
// "seen" without any I/O.
//
// # Scrub
//
// A scrub operation re-hashes a configurable percentage of files that were
// previously recorded by a sync.  Files are selected in least-recently-scrubbed
// order so that all files are eventually verified.  A hash mismatch indicates
// silent data corruption (bit rot).
//
// # Worker pool
//
// The directory walker runs as a producer goroutine that sends [domain.WorkItem]
// values into a buffered channel.  A pool of N consumer goroutines reads from
// the channel, hashes each file using [domain.Hasher], and sends
// [domain.WorkResult] values back through a second channel.  The collector
// goroutine receives results and writes them to the [domain.Repository].
//
// The worker count N is supplied by the caller and is typically determined by
// [monitor.Detect] to avoid head thrashing on spinning-platter drives.
package scanner

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/YaoxuanZhang/bit-rot-detector/internal/domain"
)

const workBufferSize = 512

// Syncer orchestrates a sync operation: it walks the directory tree
// (producer), hashes new/modified files via a worker pool (consumer), and
// writes results to the repository.
type Syncer struct {
	hasher     domain.Hasher
	numWorkers int
}

// NewSyncer creates a Syncer.
func NewSyncer(h domain.Hasher, numWorkers int) *Syncer {
	if numWorkers < 1 {
		numWorkers = 1
	}
	return &Syncer{hasher: h, numWorkers: numWorkers}
}

// SyncDirectory walks rootPath, compares each file against the repository,
// and records new, modified, moved, and deleted files.
func (s *Syncer) SyncDirectory(ctx context.Context, rootPath string, repo domain.Repository) (*domain.SyncResult, error) {
	// Respect cancellation before any work begins.
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	slog.Info("sync started", "path", rootPath, "workers", s.numWorkers)

	// Load existing records from the shadow DB.
	dbFiles, err := repo.GetAllFiles(ctx)
	if err != nil {
		return nil, fmt.Errorf("load db files: %w", err)
	}
	slog.Info("loaded existing records", "count", len(dbFiles))

	// work channel: producer -> workers
	work := make(chan domain.WorkItem, workBufferSize)
	// results channel: workers -> collector
	results := make(chan domain.WorkResult, workBufferSize)

	// Walk-phase bookkeeping (populated by walker, read by collector).
	// Maps abs_path -> (size, mtime) for every file actually seen on disk.
	var (
		seenMu  sync.Mutex
		seenFiles = make(map[string]struct{ size int64; mtime time.Time })
	)

	// Start worker pool.
	var wg sync.WaitGroup
	for i := 0; i < s.numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for item := range work {
				select {
				case <-ctx.Done():
					return
				default:
				}
				// Short-circuit: reuse the stored hash if the file is unchanged.
				if item.KnownHash != "" {
					results <- domain.WorkResult{
						Path:  item.Path,
						Hash:  item.KnownHash,
						Size:  item.Size,
						Mtime: item.Mtime,
					}
					continue
				}
				hash, herr := s.hasher.ComputeHash(ctx, item.Path)
				results <- domain.WorkResult{
					Path:  item.Path,
					Hash:  hash,
					Size:  item.Size,
					Mtime: item.Mtime,
					Err:   herr,
				}
			}
		}()
	}

	// Goroutine to close results once all workers are done.
	go func() {
		wg.Wait()
		close(results)
	}()

	result := &domain.SyncResult{}
	// resultMu protects result.Errors and all result counter fields that are
	// written from both the walker goroutine and the collector goroutine.
	var resultMu sync.Mutex

	// Producer: walk directory tree.
	walkErr := make(chan error, 1)
	go func() {
		defer close(work)
		err := filepath.WalkDir(rootPath, func(path string, d os.DirEntry, err error) error {
			if err != nil {
				slog.Warn("walk error", "path", path, "err", err)
				resultMu.Lock()
				result.Errors = append(result.Errors, err.Error())
				resultMu.Unlock()
				return nil // continue walking
			}
			if d.IsDir() {
				return skipSystemDir(d.Name())
			}
			if d.Type()&os.ModeSymlink != 0 {
				return nil // skip symlinks
			}

			// Skip the database files (including WAL and SHM files) and canary.
			base := filepath.Base(path)
			if base == "bitrot.db" || base == "bitrot.db.shadow" ||
				base == "bitrot.db-wal" || base == "bitrot.db-shm" ||
				base == "bitrot.db.shadow-wal" || base == "bitrot.db.shadow-shm" ||
				base == ".bitrot-canary" {
				return nil
			}

			info, serr := d.Info()
			if serr != nil {
				slog.Warn("stat error", "path", path, "err", serr)
				resultMu.Lock()
				result.Errors = append(result.Errors, serr.Error())
				resultMu.Unlock()
				return nil
			}

			absPath, _ := filepath.Abs(path)
			size := info.Size()
			mtime := info.ModTime()

			seenMu.Lock()
			seenFiles[absPath] = struct {
				size  int64
				mtime time.Time
			}{size, mtime}
			seenMu.Unlock()

			resultMu.Lock()
			result.FilesScanned++
			scanned := result.FilesScanned
			resultMu.Unlock()
			if scanned%1000 == 0 {
				slog.Info("walking...", "scanned", scanned)
			}

			// Short-circuit: if size and mtime match the stored record, skip
			// re-hashing entirely and reuse the stored hash.
			item := domain.WorkItem{Path: absPath, Size: size, Mtime: mtime}
			if existing, ok := dbFiles[absPath]; ok &&
				existing.FileSize == size &&
				existing.Mtime.Unix() == mtime.Unix() {
				item.KnownHash = existing.Hash
			}

			select {
			case <-ctx.Done():
				return ctx.Err()
			case work <- item:
			}
			return nil
		})
		walkErr <- err
	}()

	// Collector: process worker results and write to the repository.
	for res := range results {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		default:
		}

		if res.Err != nil {
			slog.Warn("hash error", "path", res.Path, "err", res.Err)
			resultMu.Lock()
			result.Errors = append(result.Errors, fmt.Sprintf("hash %s: %v", res.Path, res.Err))
			resultMu.Unlock()
			continue
		}

		existing, inDB := dbFiles[res.Path]
		rec := &domain.FileRecord{
			AbsPath:  res.Path,
			Hash:     res.Hash,
			FileSize: res.Size,
			Mtime:    res.Mtime,
		}

		if inDB {
			// Check whether file was actually modified.
			sizeChanged := existing.FileSize != res.Size
			mtimeChanged := existing.Mtime.Unix() != res.Mtime.Unix()

			if sizeChanged || mtimeChanged {
				resultMu.Lock()
				result.FilesModified++
				resultMu.Unlock()
				slog.Info("file modified", "path", res.Path)
			}
			rec.AddedAt = existing.AddedAt
			rec.LastScrubbed = existing.LastScrubbed
			rec.ScrubCount = existing.ScrubCount
		} else {
			// Check for a moved file (same size+mtime at a different path).
			// Protect seenFiles with the mutex since the walker goroutine writes to it concurrently.
			seenMu.Lock()
			movedFrom := findMovedFile(res.Path, res.Size, res.Mtime, dbFiles, seenFiles)
			seenMu.Unlock()
			if movedFrom != "" {
				old := dbFiles[movedFrom]
				rec.AddedAt = old.AddedAt
				rec.LastScrubbed = old.LastScrubbed
				rec.ScrubCount = old.ScrubCount
				rec.Hash = old.Hash // reuse stored hash — file content unchanged
				if err := repo.DeleteFile(ctx, movedFrom); err != nil {
					slog.Warn("delete moved-from record", "path", movedFrom, "err", err)
				}
				resultMu.Lock()
				result.FilesMoved++
				resultMu.Unlock()
				slog.Info("file moved", "from", movedFrom, "to", res.Path)
			} else {
				rec.AddedAt = time.Now()
				resultMu.Lock()
				result.FilesAdded++
				resultMu.Unlock()
				slog.Debug("file added", "path", res.Path)
			}
		}

		if err := repo.UpsertFile(ctx, rec); err != nil {
			slog.Warn("upsert file", "path", res.Path, "err", err)
			resultMu.Lock()
			result.Errors = append(result.Errors, fmt.Sprintf("upsert %s: %v", res.Path, err))
			resultMu.Unlock()
		}
	}

	// Wait for walker.
	if werr := <-walkErr; werr != nil {
		if ctx.Err() != nil {
			// Walk was stopped by context cancellation; propagate as cancellation.
			return result, ctx.Err()
		}
		return result, fmt.Errorf("walk: %w", werr)
	}

	// Phase 3: detect deleted files (in DB but not seen on disk).
	seenMu.Lock()
	defer seenMu.Unlock()
	for absPath := range dbFiles {
		if _, seen := seenFiles[absPath]; !seen {
			slog.Info("file deleted", "path", absPath)
			if err := repo.DeleteFile(ctx, absPath); err != nil {
				slog.Warn("delete missing record", "path", absPath, "err", err)
			}
			result.FilesRemoved++
		}
	}

	slog.Info("sync complete",
		"scanned", result.FilesScanned,
		"added", result.FilesAdded,
		"modified", result.FilesModified,
		"moved", result.FilesMoved,
		"removed", result.FilesRemoved,
		"errors", len(result.Errors),
	)
	return result, nil
}

// ScrubFiles re-verifies a percentage of files to detect bit rot.
func (s *Syncer) ScrubFiles(ctx context.Context, repo domain.Repository, percentage float64, minAgeDays *int) (*domain.ScrubResult, error) {
	// Respect cancellation before any work begins.
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	slog.Info("scrub started", "percentage", percentage)

	files, err := repo.GetFilesForScrub(ctx, percentage, minAgeDays)
	if err != nil {
		return nil, fmt.Errorf("get files for scrub: %w", err)
	}
	slog.Info("files selected for scrub", "count", len(files))

	res := &domain.ScrubResult{}

	for _, rec := range files {
		select {
		case <-ctx.Done():
			return res, ctx.Err()
		default:
		}

		if _, serr := os.Stat(rec.AbsPath); os.IsNotExist(serr) {
			slog.Warn("file not found during scrub", "path", rec.AbsPath)
			res.Errors = append(res.Errors, fmt.Sprintf("not found: %s", rec.AbsPath))
			continue
		}

		hash, herr := s.hasher.ComputeHash(ctx, rec.AbsPath)
		if herr != nil {
			slog.Warn("hash error during scrub", "path", rec.AbsPath, "err", herr)
			res.Errors = append(res.Errors, fmt.Sprintf("hash %s: %v", rec.AbsPath, herr))
			continue
		}

		if hash != rec.Hash {
			slog.Error("BIT ROT DETECTED", "path", rec.AbsPath,
				"expected", rec.Hash, "got", hash)
			res.FilesCorrupted = append(res.FilesCorrupted, rec.AbsPath)
			continue
		}

		if uerr := repo.UpdateScrubStatus(ctx, rec.AbsPath); uerr != nil {
			slog.Warn("update scrub status", "path", rec.AbsPath, "err", uerr)
		}
		res.FilesValidated++
		if res.FilesValidated%100 == 0 {
			slog.Info("scrub progress", "validated", res.FilesValidated, "total", len(files))
		}
	}

	slog.Info("scrub complete",
		"validated", res.FilesValidated,
		"corrupted", len(res.FilesCorrupted),
		"errors", len(res.Errors),
	)
	return res, nil
}

// skipSystemDir returns filepath.SkipDir for well-known Windows system folders
// and hidden directories to avoid unnecessary traversal.
func skipSystemDir(name string) error {
	switch name {
	case "$RECYCLE.BIN", "System Volume Information", "Recovery",
		"$Recycle.Bin", "Windows", "Program Files", "Program Files (x86)",
		"ProgramData":
		return filepath.SkipDir
	}
	return nil
}

// findMovedFile searches dbFiles for a record whose size and mtime match and
// whose old path is no longer present on disk (seenFiles).  This prevents
// false-positive move detection when duplicates exist.
func findMovedFile(
	newPath string,
	size int64,
	mtime time.Time,
	dbFiles map[string]*domain.FileRecord,
	seenFiles map[string]struct{ size int64; mtime time.Time },
) string {
	for dbPath, rec := range dbFiles {
		if dbPath == newPath {
			continue
		}
		// Old path must no longer be visible on disk.
		if _, stillExists := seenFiles[dbPath]; stillExists {
			continue
		}
		if rec.FileSize == size && rec.Mtime.Unix() == mtime.Unix() {
			return dbPath
		}
	}
	return ""
}
